cpp/src/copying/gather.cu

```diff
@@ -0,0 +1,106 @@
+#include "gather.cuh"
+#include <cudf/types.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/copying.hpp>
+#include <utilities/error_utils.hpp>
+
+#include <rmm/thrust_rmm_allocator.h>
+#include <thrust/count.h>
+
+#include <memory>
+
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+
+struct dispatch_map_type {
+  template <typename map_type, std::enable_if_t<std::is_integral<map_type>::value>* = nullptr>
+  std::unique_ptr<table> operator()(table_view const& source_table,
+				    column_view const& gather_map,
+				    size_type num_destination_rows, bool check_bounds,
+				    bool ignore_out_of_bounds,
+				    bool allow_negative_indices = false)
+  {
+    std::unique_ptr<table> destination_table;
+    map_type const * typed_gather_map = gather_map.data<map_type>();
+
+    if (check_bounds) {
+      cudf::size_type begin = (allow_negative_indices) ? -source_table.num_rows() : 0;
+      CUDF_EXPECTS(
+	  num_destination_rows == thrust::count_if(
+	      rmm::exec_policy()->on(0),
+	      typed_gather_map,
+	      typed_gather_map + num_destination_rows,
+	      bounds_checker<map_type>{begin, source_table.num_rows()}),
+	  "Index out of bounds.");
+    }
+
+    if (allow_negative_indices) {
+      destination_table =
+	gather(source_table,
+	       thrust::make_transform_iterator(
+					       typed_gather_map,
+					       index_converter<map_type,index_conversion::NEGATIVE_TO_POSITIVE>{source_table.num_rows()}),
+	       num_destination_rows,
+	       check_bounds,
+	       ignore_out_of_bounds,
+	       allow_negative_indices
+	     );
+    }
+    else {
+      destination_table =
+	gather(source_table,
+	       thrust::make_transform_iterator(
```

Comment:
Perhaps I'm missing something, but the `index_converter` seems to add unnecessary complexity.

`gather_map.begin<map_type>()` could simply be passed instead. (A raw pointer is a valid iterator).